src/dialogue.agda

```diff
@@ -20,30 +21,91 @@ X ^ω = Seq X
 -- of a functional on the points of a spread.
 data 𝓓 (Y Z : Set) : Set where
   η_ : Z → 𝓓 Y Z
-  ϝ : (Y → 𝓓 Y Z) → 𝓓 Y Z
+  ϝ : Nat → (Y → 𝓓 Y Z) → 𝓓 Y Z
+
+-- In the normalized (Brouwerian) version of the dialogue tree, queries are given in order.
+data 𝓓ₙ (Y Z : Set) : Set where
+  η_ : Z → 𝓓ₙ Y Z
+  ϝ : (Y → 𝓓ₙ Y Z) → 𝓓ₙ Y Z
+
+nth : {Y : Set} → List Y → Nat → Y ⊕ 𝟙
+nth [] i = ⊕.inr 𝟙.*
+nth (x ∷ xs) ze = ⊕.inl x
+nth (x ∷ xs) (su_ i) = nth xs i
+
+
+-- TODO: This needs to be reimplemented structurally and proved correct.
+-- My concern is that the ability to write this function is itself the
+-- content of the bar induction principle--I don't know if that is actually
+-- a correct statement, but it does look a lot like the "proof" of BI that I
+-- gave in Agda here: https://github.com/jonsterling/agda-bar-induction/blob/master/src/BarTheorem.agda#L62
+--
+{-# TERMINATING #-}
+normalize : {Y Z : Set} → List Y → 𝓓 Y Z → 𝓓ₙ Y Z
```

Comment:
@freebroccolo I am curious what you think about this function—does it seem to you that it can be implemented as a terminating function in Agda?

My worry is that the general recursive implementation I have given here looks almost exactly like the bar functional itself—i.e. its termination looks almost like it is the content of the bar principle, so we cannot use it to prove it.

However, because of the wellfounded nature of the input, it seems like we actually should be able to do implement this normalization procedure properly. What do you think?